src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -1721,12 +1765,22 @@ else if (null == currentPrimaryPlaceHolder) {
             // Update timeout interval (but no more than the point where we're supposed to fail: timerExpire)
             attemptNumber++;
 
-            if (useParallel || useTnir) {
+            if (useParallel) {
                 intervalExpire = System.currentTimeMillis() + (timeoutUnitInterval * (attemptNumber + 1));
             }
             else if (isDBMirroring) {
                 intervalExpire = System.currentTimeMillis() + (timeoutUnitInterval * ((attemptNumber / 2) + 1));
             }
+            else if (useTnir) {
+                long timeSlice = timeoutUnitInterval * ((long) Math.pow(2, attemptNumber));
```

Comment:
Instead of using `Math.pow` , I would recommend using 1 << attemptNumber 
Logical shift should be more performant.